snuba/clickhouse/http.py

```diff
@@ -73,15 +94,80 @@ def write(self, rows: Iterable[WriterTableRow]):
             chunked=True,
         )
 
-        if response.status != 200:
-            # XXX: This should be switched to just parse the JSON body after
-            # https://github.com/yandex/ClickHouse/issues/6272 is available.
-            content = response.data.decode("utf8")
-            details = CLICKHOUSE_ERROR_RE.match(content)
-            if details is not None:
-                code, type, message = details.groups()
-                raise ClickhouseError(int(code), message)
-            else:
-                raise HTTPError(
-                    f"Received unexpected {response.status} response: {content}"
-                )
+        raise_for_error_response(response)
+
+
+def parse_and_transform_date(value: str) -> str:
+    return transform_date(dateutil_parse(value))
+
+
+def parse_and_transform_datetime(value: str) -> str:
+    return transform_datetime(dateutil_parse(value))
+
+
+transform_column_types = build_result_transformer(
+    [
+        (re.compile(r"^Date(\(.+\))?$"), parse_and_transform_date),
+        (re.compile(r"^DateTime(\(.+\))?$"), parse_and_transform_datetime),
+    ]
+)
+
+
+class HTTPReader(Reader[ClickhouseQuery]):
+    def __init__(
+        self, host: str, port: int, settings: Optional[Mapping[str, str]] = None
+    ):
+        if settings is not None:
+            assert "query_id" not in settings, "query_id cannot be passed as a setting"
+
+        self.__pool = HTTPConnectionPool(host, port)
+
+        self.__default_settings: MutableMapping[str, str] = (
+            {**settings} if settings is not None else {}
+        )
+
+        if "output_format_json_quote_64bit_integers" not in self.__default_settings:
+            self.__default_settings["output_format_json_quote_64bit_integers"] = "0"
+
+    def execute(
+        self,
+        query: ClickhouseQuery,
+        settings: Optional[Mapping[str, str]] = None,
+        query_id: Optional[str] = None,
+        with_totals: bool = False,  # NOTE: unnecessary with FORMAT JSON
```

Comment:
This comment I guess is a bit misleading. `FORMAT JSON` doesn't implicitly add `WITH TOTALS` or anything, but it does already split out the `totals` row from the `data` rows so that the caller doesn't need to. When the AST is available everywhere, it'd be better to remove `with_totals` from this function signature and just use the query attribute to decide whether or not the totals row should exist or not for the native reader.